File: week1/이다혜/다리를지나는트럭.java
import java.util.LinkedList;
import java.util.Queue;

public class 프로그래머스_다리를지나는트럭 {

    public int solution(int bridge_length, int weight, int[] truck_weights) {
        
        int answer = 0; 
        int currentBridgeWeight = 0, currentTruckIdx=0;
        int lastTruckIdx = truck_weights.length;
        
        Queue<Integer> q = new LinkedList<>();           
        
        while(currentTruckIdx < lastTruckIdx){
            if(q.size() == bridge_length){
                currentBridgeWeight -= q.poll();                
            }

            else if(currentBridgeWeight + truck_weights[currentTruckIdx] > weight){
                q.offer(0);
                answer++;
            }else{
                q.offer(truck_weights[currentTruckIdx]);
                currentBridgeWeight += truck_weights[currentTruckIdx++];
                answer++;
            }            
        }

        return answer + bridge_length;
    }
    
}

File: week1/한지희/다리를 지나는 트럭.kt
import java.util.*

data class WeightDistance(val truckWeight: Int, var distance: Int)

class BridgeTruck {
    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        val readyTrucks = truck_weights.toMutableList()
        val bridgeTrucks = mutableListOf<WeightDistance>()
        var time = 0
        var currentWeight = 0
        var currentTrucksCount = 0

        while (readyTrucks.isNotEmpty() || bridgeTrucks.isNotEmpty()) {
            time++
            if (bridgeTrucks.isNotEmpty()) {
                bridgeTrucks.map { it.distance++ }
                if (bridgeTrucks[0].distance >= bridge_length) {
                    val removeTruck = bridgeTrucks.removeFirst()
                    currentWeight -= removeTruck.truckWeight
                    currentTrucksCount--
                }
            }

            if (readyTrucks.isNotEmpty()) {
                val firstTruck = readyTrucks[0]
                if (currentWeight + firstTruck <= weight && currentTrucksCount + 1 <= bridge_length) {
                    readyTrucks.removeFirst()
                    bridgeTrucks.add(WeightDistance(firstTruck, 0))
                    currentWeight += firstTruck
                    currentTrucksCount++
                }
            }
        }

        return time
    }

    // 다리의 길이만큼의 크기를 가진 큐 생성 -> 한 칸씩 빼고, 0을 넣으면서 앞으로 이동
    fun bestSolution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        var answer = 0
        val bridgeQueue: Queue<Int> = LinkedList(List(bridge_length) { 0 })
        val waitingQueue: Queue<Int> = LinkedList(truck_weights.toList())

        while (bridgeQueue.isNotEmpty()) {
            answer++
            bridgeQueue.poll()
            if (waitingQueue.isNotEmpty()) {
                if (bridgeQueue.sum() + waitingQueue.peek() <= weight) {
                    bridgeQueue.add(waitingQueue.poll())
                } else {
                    bridgeQueue.add(0)
                }
            }
        }
        return answer
    }
}